include/up-cpp/transport/datamodel/UPayload.h

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright (c) 2024 General Motors GTO LLC
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * 
+ * SPDX-FileType: SOURCE
+ * SPDX-FileCopyrightText: 2024 General Motors GTO LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef _UPAYLOAD_H_
+#define _UPAYLOAD_H_
+
+#include <memory>
+#include <cstdint>
+#include <cstring>
+
+namespace uprotocol::utransport {
+
+    enum class UPayloadType {
```

Comment:
Regarding UPayload, it's an object holding a shared_ptr, wouldn't it be more C++ idiomatic for payloads to be treated as shared_ptr's to UPayload, and for UPayload to be derived from an STL container with views support, such as std::basic_string<uint8_t> and std::basic_string_view<uint8_t>. This would allow client code to pass data in-place (as long as the data is consumed synchronously).

Reply:
> The proto file looks like a naked pointer, and those always raise concerns around data safety for me.

_Yes_, this also sounded alarm bells for me

Reply:
The current zenoh-cpp stuff is uniformly using std::string and std::string_view as generic byte containers. My preference for std::basic_string<> with either char or uint8_t is that the std::basic_string_view<> class already exists and has known properties. I think there is a more explicit guarantee that the bytes are going to be consecutive in memory. It is the case that vector<uint8_t> or whatever gives you consecutive bytes, but is that promised in writing in the STL spec? Choosing char over uint8_t for std::basic_string<> is easiest because both zenoh-cpp and protobuf regard std::string as the basic byte container. I find that aesthetically displeasing for its presumption about the meaning of the data.

Reply:
Re: pointers and the concept of the UPayload reference, I have filed https://github.com/eclipse-uprotocol/up-spec/issues/128 to discuss either removing it until we can design a better zero-copy / shared memory API _or_ how such an API should be designed with better data safety guarantees.

----

@debruce The first sentence of the [cppreference.com page for std::vector](https://en.cppreference.com/w/cpp/container/vector) states:

> The elements are stored contiguously, which means that elements can be accessed not only through iterators, but also using offsets to regular pointers to elements.

This was clarified for C++98 with defect report [LWG 69](https://cplusplus.github.io/LWG/issue69). It is a well-defined, explicit guarantee of `std::vector`. (NB: You can also check 26.3.11.1, paragraph 2 in the [C++17 spec](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2017/n4659.pdf) if cppreference.com doesn't convince you)

As for the choice between using `std::vector<uint8_t>` and `std::string` (which is `std::basic_string<char>`, and `char` is, in turn, often equivalent to `int8_t`).  IMO, `std::vector<uint8_t>` is the idiomatic way to represent a sequence of non-character bytes in C++. A `std::string` is defined as _"...stor[ing] and manipulat[ing] sequences of character-like objects..."_ - it is intended to be used in cases where the data is interpreted as a sequence of characters. By using `std::vector<uint8_t>`, it is entirely unambiguous that the data is not intended to be interpreted as a sequence of characters (i.e. you wouldn't print it to the console without some sort of conversion first).

Doing a little historical research, using `std::string` as a generic byte container appears to come from the [small string optimization](https://github.com/elliotgoodrich/SSO-23) properties of `std::string`. If many of your buffers are shorter than 16 bytes (note: this is implementation dependent), extra memory allocations can be avoided all together since the whole buffer fits inside the string object itself. **It is not possible to produce a `UMessage` shorter than 16 bytes, so we do not benefit from SSO.** (well, with gcc anyway)

With respect to protobuf using `std::string` as a byte container, we have to copy data into the protobuf objects anyway, so there is no performance gain by using `std::string` in our interface. For zenoh-cpp, it is a header-only wrapper that provides a pointer and length into the zenoh-c layer. I would imagine it is trivial to extend that interface to take either `std::string` or `std::vector<uint8_t>` with no impact to performance (assuming that is something the zenoh-cpp maintainers are open to).

Reply:
References were removed from UPayload in the spec until a true zero-copy / shared memory solution is designed, so UPayload is just a byte buffer and type now.